src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SymbolToProjectId.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        /// <summary>
+        /// Contains caches to help us quickly and accurately map from <see cref="ISymbol"/>s to <see
+        /// cref="ProjectId"/>.
+        /// </summary>
+        /// <remarks>
+        /// Note: this is not <see langword="readonly"/> as this is a struct and will mutate as operations are called on
+        /// it.  However, in the common case where this data is not needed, it will have not have any impact.
+        /// </remarks>
+        private SymbolToProjectId _symbolToProjectId;
+
+        /// <inheritdoc cref="Solution.GetExactProjectId"/>
+        public ProjectId? GetExactProjectId(ISymbol symbol)
+            => _symbolToProjectId.GetExactProjectId(symbol);
+
+        /// <summary>
+        /// A helper type that can be used to map from symbols to a project they could have come from.
+        /// </summary>
+        private struct SymbolToProjectId
+        {
+            /// <summary>
+            /// Weak cache from a <see cref="Compilation"/> to the ID of the <see cref="Project"/> it was created from.
+            /// Only used for mapping a <see cref="Project"/>'s <see cref="Compilation.GlobalNamespace"/> back.  These
+            /// <see cref="INamespaceSymbol"/>'s do not belong to a <see cref="IAssemblySymbol"/> or <see
+            /// cref="IModuleSymbol"/> and thus cannot be tracked with <see
+            /// cref="CompilationTracker.State.CompilationAssembliesAndModules"/>.
+            /// </summary>
+            private static readonly ConditionalWeakTable<Compilation, ProjectId> s_compilationToProjectId
+                = new ConditionalWeakTable<Compilation, ProjectId>();
```

Comment:
So nothing shares this from one solution snapshot to the next? CompilationTrackers can be reused between Solution snapshots if a change to a Solution doesn't impact unrelated projects. It seems like an edit such as that would blow away this map so future queries won't work correctly?

Reply:
this is a static map.  so it is always shared.  the one below is an instance map.

Reply:
Should this map just go next to s_assemblyOrModuleSymbolToProjectMap then?

Reply:
to answer this question: no.

These are very different maps (though they seem similar).  The existing one maps source assembly/module symbols to a the project for them.  The new one maps from Compilations to Projects *specifically* for the case of Global-Compilation-Namespaces.  

That said, i belive we can go from compilation-namespaces to source-module-namespaces.  So i think i may be able to remove the map entirely :)

Reply:
Yaay.  Was able to remove this entirely.  This means there's only one piece of state we need here.